wordpress-mobile/WordPress-Android
# Enable submitting a custom vertical on the Site Intent Question screen

Closes [#16127](https://github.com/wordpress-mobile/WordPress-Android/issues/16127)

## Description

This PR adds the search input to the Site Creation flow, together with partial functionality and UX related to the input.
- Transition animations were also added when scrolling down / up, and
- The header section is removed in landscape mode, where we're showing the "Site Intent" title in the app bar instead.

---

### Progress Status
The following checks from the issue acceptance criteria & other considerations are implemented:

- [x] Search input is displayed above the verticals list
  - [x] Placeholder with examples is displayed inside search input
- [x] Search input tap (focus)
  - [x] hides Title & Subtitle in Header (with parallax slide down transition)
- [x] Search input typing:
  - [x] Blank white surface is displayed where the search results are supposed to appear
  - [x] Continue button shows above keyboard
    - [x] Continue button tap → Navigates to "Choose a Design" screen
  - [x] Search input empty:
    - [x] Placeholder with examples is displayed inside search input
    - [ ] List of default verticals is displayed
    - [x] Continue button is hidden

#### Other Considerations:

- [x] Unit Tests are added where applicable
- [x] Events are tracked:
   - [x] Check the [**Search focused** testing steps](https://github.com/wordpress-mobile/WordPress-Android/pull/16043)
   - [x] Check the [**Continue pressed** testing steps](https://github.com/wordpress-mobile/WordPress-Android/pull/16043)

## To Test

#### Prerequisites

The Site Intent Question screen is behind a feature flag (named `SITE_INTENT_QUESTION`), so we need to make sure the feature flag is **enabled** before testing the changes introduced with this PR.

<details>
   <summary>How to toggle the feature flag</summary>

   1. From `My Site` Go to `Me` → `App Settings` → `Debug settings`
   2. Scroll to `Features in development` section and tap on `SiteIntentQuestionFeatureConfig`
   3. Tap `RESTART THE APP` button
</details>

### ⓵ Test the Search Input & Scrolling Behavior

1. Launch the app
3. From `My Site` screen tap the `⋁` icon next to the site title
4. On the `Choose a site` screen tap on `➕` menu button
5. Select `Create WordPress.com site`
6. On the `What's your website about?` screen scroll down until the header title is no longer visible
   - **Verify** that the "Site topic" title appears in the app bar and remains visible until you scroll up, when the header title should be visible again (*only one title should be visible at a given moment, either in the header section or the appbar*)
7. Tap on the search input
   - **Verify** that the header section is hiding with a transition animation and the title "Site topic" in the app bar is visible
8. Type something random in the search input
   - **Verify** that the continue button appears
9. Remove all text from the search input
   - **Verify** that the continue button disappears
9. Type something again in the search input and press `Continue`
   - **Verify** that you land on the `Choose a design` screen

**Repeat** the steps above with different exploratory variations.

### ⓶ Test Events Tracking

1. Check the [**Search focused** testing steps](https://github.com/wordpress-mobile/WordPress-Android/pull/16043)
2. Check the [**Continue pressed** testing steps](https://github.com/wordpress-mobile/WordPress-Android/pull/16043)

<details>
<summary><h2>Preview</h2></summary>

https://user-images.githubusercontent.com/4588074/159789667-26df1feb-a624-4a9d-a088-3d2fe1f01c41.mov
</details>

## Regression Notes

1. Potential unintended areas of impact
   - None, these are changes only to the new screen which is behind a feature flag.
2. What I did to test those areas of impact (or what existing automated tests I relied on)
   - Manual Testing
   - I added automated tests for the implementation
3. What automated tests I added (or what prevented me from doing so)
   - Added unit tests in SiteCreationIntentsViewModelTest

### PR submission checklist:


### Known Issues:
1. The animation when the input is focused could be improved
   It's a bit jumpy, but I spent a lot of time on it and couldn't improve it until I decided I have to move forward without spending more time for it.
2. There is no transition animation for showing/hiding the `Continue` button.
3. If the input is focused, the keyboard doesn't hide when navigating to `Choose a design` screen.

## PR reviewer notes:
- Next two weeks I won't have bandwidth to focus on this PR, so I'd like to keep the scope limited to what was planned in the initial issue: [Enable submitting a custom vertical #16127](https://github.com/wordpress-mobile/WordPress-Android/issues/16127).

- Everything else about the search functionality is specified in the parent issue: [Enable searching through default verticals on the Site Intent Question screen #16053](https://github.com/wordpress-mobile/WordPress-Android/issues/16053), but we should keep in mind there are a few specs that changed internally here: pbArwn-447-p2

- I didn't add unit tests for the `onSearchTextChanged` method in ViewModel because the implementation is only temporary, and it should change when #16053 is implemented.
